wgplugin.cpp

```diff
@@ -0,0 +1,231 @@
+{
+   return true;
+}
+
+bool WGPlugin::parse_wg(const char *data, unsigned int payload_len, bool source_pkt, RecordExtWG *ext)
+{
+   uint32_t cmp_peer;
+   uint32_t cmp_new_peer;
+
+   total++;
+
+   // The smallest message (according to specs) is the data message (0x04) with 16 header bytes
+   // and 16 bytes of (empty) data authentication.
+   // Anything below that is not a valid WireGuard message.
+   if (payload_len < WG_PACKETLEN_MIN_TRANSPORT_DATA) {
+      return false;
+   }
+
+   // Let's try to parse according to the first 4 bytes, and see if that is enough.
+   // The first byte is 0x01-0x04, the following three bytes are reserved (0x00).
+   uint8_t pkt_type = data[0];
+   if (pkt_type < WG_PACKETTYPE_INIT_TO_RESP || pkt_type > WG_PACKETTYPE_TRANSPORT_DATA) {
+      return false;
+   }
+   if (data[1] != 0x0 || data[2] != 0x0 || data[3] != 0x0) {
+      return false;
+   }
+
+   // Next, check the packet contents based on the message type.
+   switch(pkt_type) {
+      case WG_PACKETTYPE_INIT_TO_RESP:
+         if (payload_len != WG_PACKETLEN_INIT_TO_RESP) {
+            return false;
+         }
+         
+         // compare the current dst_peer and see if it matches the original source.
+         // If not, the flow flush may be needed to create a new flow.
+         cmp_peer = source_pkt ? ext->src_peer : ext->dst_peer;
+         memcpy(&cmp_new_peer, (data+4), sizeof(uint32_t));
+         
+         if (cmp_peer != 0 && cmp_peer != cmp_new_peer) {
+            flow_flush = true;
+            return false;
+         }
+
+         memcpy(source_pkt ? &(ext->src_peer) : &(ext->dst_peer), (data+4), sizeof(uint32_t));
+         break;
+
+      case WG_PACKETTYPE_RESP_TO_INIT:
+         if (payload_len != WG_PACKETLEN_RESP_TO_INIT) {
+            return false;
+         }
+
+         memcpy(&(ext->src_peer), (data+4), sizeof(uint32_t));
+         memcpy(&(ext->dst_peer), (data+8), sizeof(uint32_t));
+         
+         // let's swap for the opposite direction
+         if (! source_pkt) {
+            swap(ext->src_peer, ext->dst_peer);
+         }
+         break;
+   
+      case WG_PACKETTYPE_COOKIE_REPLY:
+         if (payload_len != WG_PACKETLEN_COOKIE_REPLY) {
+            return false;
+         }
+
+         memcpy(source_pkt ? &(ext->dst_peer) : &(ext->src_peer), (data+4), sizeof(uint32_t));
+         break;
+
+      case WG_PACKETTYPE_TRANSPORT_DATA:
+         // Each packet of transport data is zero-padded to the multiple of 16 bytes in length.
+         if (payload_len < WG_PACKETLEN_MIN_TRANSPORT_DATA || (payload_len % 16) != 0) {
+            return false;
+         }
+
+         memcpy(source_pkt ? &(ext->dst_peer) : &(ext->src_peer), (data+4), sizeof(uint32_t));
+         break;
+   }
+
+   ext->possible_wg = 100;
+   identified++;
+   return true;
+}
+
+int WGPlugin::add_ext_wg(const char *data, unsigned int payload_len, bool source_pkt, Flow &rec)
+{
+   RecordExtWG *ext = new RecordExtWG();
```

Comment:
There is a memory allocation each time UDP packet arrives. Better solution is to preallocate record.

Reply:
OK, got it now, should be fixed.